src/file-manager/fm-desktop-wayland-bg-dialog.c

```diff
@@ -0,0 +1,454 @@
+                                          "color-shading-type");
+
+    secondary_color_str = g_settings_get_string (background_settings,
+                                               "secondary-color");
+
+    gtk_color_chooser_get_rgba (GTK_COLOR_CHOOSER (colorbutton1), &color1);
+    primary_color_str = gdk_rgba_to_string (&color1);
+
+    update_preview (FALSE, NULL, "", shading_type,
+                    primary_color_str, secondary_color_str);
+
+    g_settings_set_string (background_settings,
+                           "primary-color", primary_color_str);
+
+    g_settings_set_string (background_settings,
+                       "picture-filename", "");
+}
+
+static void
+update_secondary_color (GtkWidget *colorbutton2)
+{
+    const gchar *shading_type, *primary_color_str, *secondary_color_str;
+    GdkRGBA color2;
+
+    shading_type = g_settings_get_string (background_settings,
+                                          "color-shading-type");
+
+    primary_color_str = g_settings_get_string (background_settings,
+                                               "primary-color");
+
+    gtk_color_chooser_get_rgba (GTK_COLOR_CHOOSER (colorbutton2), &color2);
+    secondary_color_str = gdk_rgba_to_string (&color2);
+
+    g_settings_set_string (background_settings,
+                           "secondary-color", secondary_color_str);
+
+    g_settings_set_string (background_settings,
+                       "picture-filename", "");
+
+    update_preview (FALSE, NULL, "", shading_type,
+                    primary_color_str, secondary_color_str);
+
+}
+
+static void
+update_color_background_options (GtkWidget *colorbox)
+{
+    const gchar *shading_type, *primary_color_str, *secondary_color_str;
+
+    primary_color_str = g_settings_get_string (background_settings,
+                                               "primary-color");
+
+    secondary_color_str = g_settings_get_string (background_settings,
+                                               "secondary-color");
+
+    shading_type = gtk_combo_box_get_active_id (GTK_COMBO_BOX (colorbox));
+    /*write to gsettings*/
+    g_settings_set_string (background_settings,
+                           "color-shading-type", shading_type);
+    g_settings_set_string (background_settings,
+                       "picture-filename", "");
+
+    update_preview (FALSE, NULL, "", shading_type,
+                    primary_color_str, secondary_color_str);
+}
+
+static void
+update_image_background_options(GtkWidget *stylebox)
+{
+    const gchar *options;
+    options = gtk_combo_box_get_active_id (GTK_COMBO_BOX(stylebox));
+
+    /*write to gsettings*/
+    g_settings_set_string (background_settings,
+                           "picture-options", options);
+
+    /*Only the image changes here, we are not thumbnailing image options yet*/
+}
+
+static void
+update_background_image (GtkWidget *filebutton)
+{
+    const gchar *filename = gtk_file_chooser_get_filename (GTK_FILE_CHOOSER (filebutton));
```

Comment:
Memory should be released ```g_free (filename)```

Reply:
New line 263 for this is
`    g_free (filename);`